### ISSUE TITLE
Make sure that passing in NaN weights do not cause a crash

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1560]
Passing NaNs as weights will cause some functions to crash, see e.g. https://github.com/igraph/rigraph/issues/415

Functions which can't handle NaN should check the weight vector for NaNs.

Note that some functions may only access a small part of the weight vector in some cases. Ideally, they should not loop over the entire weight vector just to check for NaNs then.

----

List of functions that are likely affected (please amend this list):

 - Anything based on shortest paths, such as betweenness, closeness, efficiency, etc.

**Comment by ntamas** [comment]
This issue is a bit vague in the sense that it's hard to tell whether the goal outlined in the title has been achieved or not (cf. the [halting problem](https://www.wikiwand.com/en/Halting_problem)). Maybe we should come up with an explicit list of functions to check/test (the ones that we think might be affected by this problem) so we know that the issue can be closed when we have gone through that list and confirmed that all of these functions work correctly.

**Comment by szhorvat** [comment]
Yes, the intention was to put a list with checkboxes in the top post. Part of the task is to identify affected functions.

**Comment by vtraag** [comment]
Shortest path functions in principle work OK with `NaN` values. That is, they do not crash. However, they are not consistent in their output and handling of `NAN`. I use a simple graph (0 - 1 - 2 - 3 - 0) where the last edge has weight NAN and the others have weight 1.

The results are as follows:

```
igraph_shortest_paths_bellman_ford
0 1 2 3
1 0 1 2
2 1 0 1
3 2 1 0
```

It is debatable whether this answer is correct. If the edge 2-3 would also be `NAN` `igraph_shortest_paths_bellman_ford` would return infinity, and says it could not be reached.

```
igraph_shortest_paths_dijkstra
0 1 NAN NAN
1 0 1 2
NAN 1 0 1
NAN 2 1 0
```

```
igraph_shortest_paths_johnson
0 1 NAN NAN
1 0 1 2
NAN 1 0 1
NAN 2 1 0
```

These two answers are correct, for the `NAN` entries here we cannot determine whether the alternative path would be shorter or not. For the provided distances, the alternative path is known to be longer. I would propose that we also need to output `NAN` in the indicated locations for `igraph_shortest_paths_bellman_ford`.

When getting (all) shortest paths to all other vertices from 0 the results are as follows.

```
igraph_get_shortest_paths_dijkstra
0
0 1
0 3 2
0 3
```

```
igraph_get_all_shortest_paths_dijkstra
0
0 1
0 3 2
0 3
```

Again, this is wrong, the shortest path to node 2 and 3 are unknown (as correctly indicated by the `NAN` in `igraph_shortest_paths_dijkstra` and `igraph_shortest_paths_johnson`). I think this should be corrected.

This issue is only for tracking whether a function crashes. Should we open a separate issue for ensuring that results are in fact *correct* when passing in NaN weights?

**Comment by vtraag** [comment]
OK, I now identified some problems. `igraph_vector_min` (and presumably some other functions) will return `nan`. Hence, checks on positivity for example do not work correctly when `nan`'s are involved.

There is then a question on what counts as valid input. Should we allow positive weights with occasional `nan`? Or should we not allow `nan` at all? As you can see from the results above, the algorithms in principle work "correctly" (i.e. they do not crash) with `nan`. Perhaps there are some use cases where it is convenient to work with `nan` for missing data, and that you can still calculate some shortest paths (if they are correct). From this point of view, it might make most sense to allow `nan` values.

**Comment by szhorvat** [comment]
Please don't make changes to `vector_min` or `vector_min`'s usage until the harmonic centrality PR is merged.

**Comment by vtraag** [comment]
No, I think that making changes to `igraph_vector_min` would not be the right direction. There are some other ways of checking the negative entries though, perhaps doing it manually. I am not sure yet of the best approach. For now the question is, do we want to allow `nan` values or not?

**Comment by ntamas** [comment]
@vtraag Please update the checklist above now that #1617 is merged. Thanks!

**Comment by vtraag** [comment]
Note to self: some other vector functions should also be checked still. Most notably `igraph_vector_(which)_minmax`, but perhaps also some others that should work correctly for NaN input still.

**Comment by szhorvat** [comment]
@vtraag Would you have time to make the list in the original post granular (one box for each function) to have a clear picture of what's missing? Many more cases have been fixed since this issue was open (I fixed several), but unfortunately we did not keep track.

**Comment by EugeniyaP** [comment]
> Yes, the intention was to put a list with checkboxes in the top post. Part of the task is to identify affected functions.

quoting the reply, comment

**Comment by szhorvat** [comment]
@EugeniyaP Was this comment made in error, or did you mean to ask something about this issue?

**Comment by EugeniyaP** [comment]
hi @szhorvat , I am exploring the possibilites of commenting on Github, there are three possibilities over all:

1) leave a comment in the end of the thread
2) leave the comment quoting someone's reply
3) leave a comment mentioning to whom you're addressing it - using @ symbol

There is no possibility to create a direct link to the comment in the thread, correct?

**Comment by szhorvat** [comment]
> There is no possibility to create a direct link to the comment in the thread, correct?

Yes, there is. Use the `...` menu at the top-right of any comment.  Example: https://github.com/igraph/igraph/issues/1560#issuecomment-1409264666

> there are three possibilities over all:

You can include anything you like in a comment, using Markdown notation. Here are the docs on that: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax  Quotation is just a special feature, it's just one of the markups possible in Markdown.

----

Let's not overload existing issues with unrelated discussions though. If you'd like to experiment, you can create a repository for this purpose under your own account. You will have full control over it and you will be able to delete comments, or even the entire repository, when you're done.

**Comment by EugeniyaP** [comment]
thank you, @szhorvat ! can I message you directly anywhere?

**Comment by szhorvat** [comment]
Yes, if you use `@` followed by my username, I will get a notification (unless I turn that off explicitly). If you need help, feel free to try this and ping me from the sandbox repository you create for the purpose of experimentation.

**Comment by EugeniyaP** [comment]
thank you!

**Comment by EugeniyaP** [comment]
thank you, @szhorvat!

**Comment by szhorvat** [comment]
> This issue is only for tracking whether a function crashes. Should we open a separate issue for ensuring that results are in fact _correct_ when passing in NaN weights?

IMO it's easier to track everything together.

I made the list at the top a bit more granular, but it's still not complete.
